tls/src/lib.rs

```diff
@@ -160,6 +204,55 @@ pub fn load_root_cert(path: &Utf8PathBuf) -> Result<Certificate, Error> {
     Ok(root)
 }
 
+fn certs_to_der(certs: &Vec<Certificate>) -> Result<Vec<u8>, Error> {
+    let mut der = Vec::new();
+
+    for cert in certs {
+        der.append(&mut cert.to_der()?);
+    }
+
+    Ok(der)
+}
+
+fn certs_from_der(buf: &[u8]) -> Result<Vec<Certificate>, Error> {
+    let mut certs = Vec::new();
+    let mut reader = SliceReader::new(buf)?;
+
+    while !reader.is_finished() {
+        certs.push(reader.decode()?);
+    }
+
+    Ok(certs)
+}
+
+async fn recv_msg<T: AsyncReadExt + Unpin>(
+    stream: &mut T,
+) -> Result<Vec<u8>, Error> {
+    // to receive a message we first get its length that is a u32 serialized as
+    // a little endian byte array
+    let mut msg_len = [0u8; 4];
+    stream.read_exact(&mut msg_len).await?;
+    let msg_len = u32::from_le_bytes(msg_len).try_into()?;
+
+    // with the length we can then get the message body
+    let mut buf = vec![0u8; msg_len];
+    stream.read_exact(&mut buf).await?;
+
+    Ok(buf)
+}
+
+async fn send_msg<T: AsyncWriteExt + Unpin>(
+    stream: &mut T,
+    msg: &[u8],
+) -> Result<(), Error> {
+    // to send a message we first send the receiver its length as a u32
+    // serialized as a little endian byte array
+    let len: u32 = msg.len().try_into()?;
+    stream.write_all(&len.to_le_bytes()).await?;
+    // then we send the message
+    Ok(stream.write_all(msg).await?)
+}
```

Comment:
This is probably a questions for @andrewjstone who has done much more with this than me but it feels like this wants to be a structure instead of just a len + bytes?